numpy/core/src/multiarray/ctors.c

```diff
@@ -3049,12 +3065,27 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
         }
         return -1;
     }
+
+    zero = PyInt_FromLong(0);
+    if (!zero) {
+        Py_DECREF(*next);
+        *next = NULL;
+        return -1;
+    }
+
+    next_is_nonzero = PyObject_RichCompareBool(*next, zero, Py_NE);
     val = PyNumber_TrueDivide(*next, step);
     Py_DECREF(*next);
     *next = NULL;
+
     if (!val) {
+        Py_DECREF(zero);
         return -1;
     }
+
+    val_is_zero = PyObject_RichCompareBool(val, zero, Py_EQ);
+    Py_DECREF(zero);
+
     if (cmplx && PyComplex_Check(val)) {
```

Comment:
Trying to work out if this section needs a fix too, but I can't work out what this is even trying to do.

Reply:
@eric-wieser Me too. I don't know why the complex case exists. This is none-sense as math.

Reply:
See #10332.